tools/gfx/renderer-shared.cpp

```diff
@@ -325,6 +331,117 @@ void PipelineStateBase::initializeBase(const PipelineStateDesc& inDesc)
     }
 }
 
+void updateCacheEntry(ISlangMutableFileSystem* fileSystem, slang::IBlob* compiledCode, String shaderFilename, slang::Checksum ASTHash)
+{
+    auto hashSize = sizeof(slang::Checksum);
+
+    auto bufferSize = hashSize + compiledCode->getBufferSize();
+    List<uint8_t> contents;
+    contents.setCount(bufferSize);
+    uint8_t* buffer = contents.begin();
+    memcpy(buffer, &ASTHash, hashSize);
+    memcpy(buffer + hashSize, (void*)compiledCode->getBufferPointer(), compiledCode->getBufferSize());
+    fileSystem->saveFile(shaderFilename.getBuffer(), buffer, bufferSize);
+}
+
+Result RendererBase::getEntryPointCodeFromShaderCache(
+    slang::IComponentType* program,
+    SlangInt entryPointIndex,
+    SlangInt targetIndex,
+    slang::IBlob** outCode,
+    slang::IBlob** outDiagnostics)
+{
+    // TODO: Need a way in filesystem to query both file size and file creation time, if cache size exceeds
+    // specified maximum size (in bytes or files) then delete oldest files - cache eviction policy
+
+    // Immediately call getEntryPointCode if no shader cache was provided on initialization
+    if (!shaderCacheFileSystem)
+    {
+        return program->getEntryPointCode(entryPointIndex, targetIndex, outCode, outDiagnostics);
+    }
+
+    // Produce a string which we can use to query the shader cache by combining two separate hashes which
+    // together comprise all the compilation arguments for this program.
+    ComPtr<slang::ISession> session;
+    getSlangSession(session.writeRef());
+
+    slang::Checksum linkageHash;
+    slang::Checksum programHash;
+    session->computeDependencyBasedHash(targetIndex, &linkageHash);
+    program->computeDependencyBasedHash(&programHash);
+
+    StringBuilder shaderFilename = checksumToString(combineChecksum(linkageHash, programHash));
+
+    // Produce a hash using the AST for this program - This is needed to check whether a cache entry is effectively dirty,
+    // or to save along with the compiled code into an entry so the entry can be checked if fetched later on.
+    slang::Checksum ASTHash;
+    program->computeASTBasedHash(&ASTHash);
+    
+    ComPtr<ISlangBlob> codeBlob;
+
+    // Query shaderCacheFileSystem for an entry whose key matches shaderFilename
+    //    - If we find it, then copy the file contents into memory and return in outCode
+    auto result = shaderCacheFileSystem->loadFile(shaderFilename.getBuffer(), codeBlob.writeRef());
+    
+    if (SLANG_FAILED(result))
+    {
+        // If we didn't find it, call program->getEntryPointCode() to get and return the code. We also
+        // make sure to save a new entry in the shader cache.
+        program->getEntryPointCode(entryPointIndex, targetIndex, codeBlob.writeRef(), outDiagnostics);
+        ISlangMutableFileSystem* fileSystem;
```

Comment:
This needs to be a `ComPtr<ISlangMutableFileSystem>` and then use `writeRef()` in the call to `queryInterface` to prevent memory leak.

Reply:
done, added `ComPtr<ISlangMutableFileSystem> mutableShaderCacheFileSystem` to `RendererBase` and moved `queryInterface` inside `RendererBase::initialize()`.